tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -3580,45 +3619,90 @@ private[optimizer] object OptimizerCore {
     val tpe: RefinedType
 
     def contains(localDef: LocalDef): Boolean = this match {
-      case PreTransBlock(_, result) =>
-        result.contains(localDef)
+      case PreTransBlock(bindingsAndStats, result) =>
+        result.contains(localDef) || bindingsAndStats.exists {
+          case Left(PreTransBinding(_, value)) => value.contains(localDef)
+          case Right(_)                        => false
+        }
       case PreTransLocalDef(thisLocalDef) =>
         thisLocalDef.contains(localDef)
       case _ =>
         false
     }
   }
 
-  private final class PreTransBlock private (val stats: List[Tree],
+  /** A pretransformed binding, part of a [[PreTransBlock]].
+   *
+   *  Even though it is not encoded in the type system, `localDef.replacement`
+   *  must be a [[ReplaceWithVarRef]] or a [[ReplaceWithRecordVarRef]].
+   */
+  private final case class PreTransBinding(localDef: LocalDef,
+      value: PreTransform) {
+
+    assert(
+        localDef.replacement.isInstanceOf[ReplaceWithVarRef] ||
+        localDef.replacement.isInstanceOf[ReplaceWithRecordVarRef],
+        "Cannot create a PreTransBinding with non-var-ref replacement " +
+        localDef.replacement)
+  }
+
+  private type BindingOrStat = Either[PreTransBinding, Tree]
+
+  private final class PreTransBlock private (
+      val bindingsAndStats: List[BindingOrStat],
       val result: PreTransLocalDef) extends PreTransform {
     def pos = result.pos
     val tpe = result.tpe
 
-    assert(stats.nonEmpty)
+    assert(bindingsAndStats.nonEmpty)
 
     override def toString(): String =
-      s"PreTransBlock($stats,$result)"
+      s"PreTransBlock($bindingsAndStats,$result)"
   }
 
   private object PreTransBlock {
-    def apply(stats: List[Tree], result: PreTransform): PreTransform = {
-      if (stats.isEmpty) result
+    def apply(bindingsAndStats: List[BindingOrStat],
+        result: PreTransLocalDef): PreTransform = {
+      if (bindingsAndStats.isEmpty) result
+      else new PreTransBlock(bindingsAndStats, result)
+    }
+
+    def apply(bindingsAndStats: List[BindingOrStat],
+        result: PreTransBlock): PreTransform = {
+      new PreTransBlock(bindingsAndStats ::: result.bindingsAndStats,
+          result.result)
+    }
+
+    def apply(binding: PreTransBinding,
+        result: PreTransLocalDef): PreTransform = {
+      if (result.contains(binding.localDef))
+        new PreTransBlock(Left(binding) :: Nil, result)
+      else
+        result // eager dce of the binding to keep the `PreTransLocalDef` nature
```

Comment:
I mean that the result keeps being a `PreTransLocalDef`, as opposed to a `PreTransBlock(..., PreTransLocalDef(...))`. Simple `PreTransLocalDef`s are better optimized in some cases than `PreTransBlock(..., PreTransLocalDef(...))` (basically because they can be duplicated everywhere we need them without risking duplicating the evaluations of the `bindingsAndStats` of the `PreTransBlock`).